# **Data Cleaning I - Investment Requirement Analysis**

Let's start with getting the datafiles rounds.csv and companies.txt.

In [1]:
from google.colab import files
uploaded = files.upload()

Saving rounds2.csv to rounds2.csv


In [2]:
from google.colab import files
uploaded = files.upload()

Saving companies.txt to companies.txt


In [3]:
import numpy as np
import pandas as pd 
import matplotlib.pyplot as plt
import seaborn as sns

# reading data files
# using encoding = "ISO-8859-1" to avoid pandas encoding error
rounds = pd.read_csv("rounds2.csv", encoding = "ISO-8859-1")
companies = pd.read_csv("companies.txt", sep="\t", encoding = "ISO-8859-1")

In [4]:
# Look at rounds head
print(rounds.head())

                     company_permalink  ... raised_amount_usd
0                  /organization/-fame  ...        10000000.0
1               /ORGANIZATION/-QOUNTER  ...               NaN
2               /organization/-qounter  ...          700000.0
3  /ORGANIZATION/-THE-ONE-OF-THEM-INC-  ...         3406878.0
4                /organization/0-6-com  ...         2000000.0

[5 rows x 6 columns]


In [5]:
# inspect the structure etc.
print(rounds.info(), "\n")
print(rounds.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 114949 entries, 0 to 114948
Data columns (total 6 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   company_permalink        114949 non-null  object 
 1   funding_round_permalink  114949 non-null  object 
 2   funding_round_type       114949 non-null  object 
 3   funding_round_code       31140 non-null   object 
 4   funded_at                114949 non-null  object 
 5   raised_amount_usd        94959 non-null   float64
dtypes: float64(1), object(5)
memory usage: 5.3+ MB
None 

(114949, 6)


In [6]:
# look at companies head
companies.head()

permalink  ...  founded_at
0                  /Organization/-Fame  ...         NaN
1               /Organization/-Qounter  ...  04-09-2014
2  /Organization/-The-One-Of-Them-Inc-  ...         NaN
3                /Organization/0-6-Com  ...  01-01-2007
4       /Organization/004-Technologies  ...  01-01-2010

[5 rows x 10 columns]

In [7]:
# companies structure
companies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66368 entries, 0 to 66367
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   permalink      66368 non-null  object
 1   name           66367 non-null  object
 2   homepage_url   61310 non-null  object
 3   category_list  63220 non-null  object
 4   status         66368 non-null  object
 5   country_code   59410 non-null  object
 6   state_code     57821 non-null  object
 7   region         58338 non-null  object
 8   city           58340 non-null  object
 9   founded_at     51147 non-null  object
dtypes: object(10)
memory usage: 5.1+ MB


Ideally, the ```permalink``` column in the companies dataframe should be the unique_key of the table, having 66368 unique company names (links, or permalinks). Also, these 66368 companies should be present in the rounds file.

Let's first confirm that these 66368 permalinks (which are the URL paths of companies' websites) are not repeating in the column, i.e. they are unique.

In [8]:
# identify the unique number of permalinks in companies
len(companies.permalink.unique())

66368

Also, let's convert all the entries to lowercase (or uppercase) for uniformity.

In [9]:
# converting all permalinks to lowercase
companies['permalink'] = companies['permalink'].str.lower()
companies.head()

permalink  ...  founded_at
0                  /organization/-fame  ...         NaN
1               /organization/-qounter  ...  04-09-2014
2  /organization/-the-one-of-them-inc-  ...         NaN
3                /organization/0-6-com  ...  01-01-2007
4       /organization/004-technologies  ...  01-01-2010

[5 rows x 10 columns]

In [10]:
# look at unique values again
len(companies.permalink.unique())

66368

Thus, there are 66368 unique companies in the table and ```permalink``` is the unique primary key. Each row represents a unique company.

Let's now check whether all of these 66368 companies are present in the rounds file, and if some extra ones are present.

In [11]:
# look at unique company names in rounds df
# note that the column name in rounds file is different (company_permalink)
len(rounds.company_permalink.unique())

90247

There seem to be 90247 unique values of ```company_permalink```, whereas we expected only 66368. May be this is because of uppercase/lowercase issues.

Let's convert the column to lowercase and look at unique values again.

In [12]:
# converting column to lowercase
rounds['company_permalink'] = rounds['company_permalink'].str.lower()
rounds.head()

company_permalink  ... raised_amount_usd
0                  /organization/-fame  ...        10000000.0
1               /organization/-qounter  ...               NaN
2               /organization/-qounter  ...          700000.0
3  /organization/-the-one-of-them-inc-  ...         3406878.0
4                /organization/0-6-com  ...         2000000.0

[5 rows x 6 columns]

In [13]:
# Look at unique values again
len(rounds.company_permalink.unique())

66370

There seem to be 2 extra permalinks in the rounds file which are not present in the companies file. Let's hope that this is a data quality issue, since if this were genuine, we have two companies whose investment round details are available but their metadata (company name, sector etc.) is not available in the companies table.


Let's have a look at the company permalinks which are in the 'rounds' file but not in 'companies'.

In [14]:
# companies present in rounds file but not in (~) companies file
rounds.loc[~rounds['company_permalink'].isin(companies['permalink']), :]

company_permalink  ... raised_amount_usd
29597                             /organization/e-cãŠbica  ...               NaN
31863        /organization/energystone-games-灵石游戏  ...               NaN
45176                /organization/huizuche-com-惠ç§ÿ车  ...               NaN
58473              /organization/magnet-tech-磁石科技  ...         1625585.0
101036  /organization/tipcat-interactive-沙èˆÿ信息ç...  ...               NaN
109969             /organization/weiche-tech-喂车科技  ...               NaN
113839                 /organization/zengame-ç¦
游科技  ...               NaN

[7 rows x 6 columns]

All the permalinks have weird non-English characters. Let's see whether these characters are present in the original df as well. 

In [15]:
# looking at the indices with weird characters
rounds_original = pd.read_csv("rounds2.csv", encoding = "ISO-8859-1")
rounds_original.iloc[[29597, 31863, 45176, 58473], :]

company_permalink  ... raised_amount_usd
29597                       /ORGANIZATION/E-CÊBICA  ...               NaN
31863  /ORGANIZATION/ENERGYSTONE-GAMES-ǁµÇŸ³Æ¸¸Æˆ  ...               NaN
45176          /organization/huizuche-com-惠ç§ÿ车  ...               NaN
58473        /ORGANIZATION/MAGNET-TECH-Ç£ÇŸ³Ç§‘ÆŠ€  ...         1625585.0

[4 rows x 6 columns]

The company weird characters appear when you import the data file. To confirm whether these characters are actually present in the given data or whether python has introduced them while importing into pandas, let's have a look at the original CSV file in Excel.

Thus, this is most likely a data quality issue we have introduced while reading the data file into python. Specifically, this is most likely caused because of encoding.

First, let's try to figure out the encoding type of this file. Then we can try specifying the encoding type at the time of reading the file. The ```chardet``` library shows the encoding type of a file.

In [16]:
# import chardet

# rawdata = open('rounds2.csv', 'rb').read()
# result = chardet.detect(rawdata)
# charenc = result['encoding']
# print(charenc)

In [17]:
# print(result)

Now let's try telling pandas (at the time of importing) the encoding type. Here's a list of various encoding types python can handle: https://docs.python.org/2/library/codecs.html#standard-encodings.

In [18]:
# trying different encodings
# encoding="cp1254" throws an error
# rounds_original = pd.read_csv("rounds2.csv", encoding="cp1254")
# rounds_original.iloc[[29597, 31863, 45176], :]

Apparently, pandas cannot decode "cp1254" in this case. After searching a lot on stackoverflow and Google, the best conclusion that can be drawn is that this file is encoded using multiple encoding types (may be because the ```company_permalink``` column contains names of companies in various countries, and hence various languages).

After trying various other encoding types (in vain), this answer suggested an alternate (and a more intelligent) way: https://stackoverflow.com/questions/45871731/removing-special-characters-in-a-pandas-dataframe.

In [19]:
rounds['company_permalink'] = rounds.company_permalink.str.encode('utf-8').str.decode('ascii', 'ignore')
rounds.loc[~rounds['company_permalink'].isin(companies['permalink']), :]

company_permalink  ... raised_amount_usd
77                                  /organization/10north  ...               NaN
729                               /organization/51wofang-  ...         5000000.0
2670                              /organization/adslinked  ...          100000.0
3166    /organization/aesthetic-everything-social-network  ...               NaN
3291                 /organization/affluent-attach-club-2  ...          100000.0
...                                                   ...  ...               ...
110545                       /organization/whodats-spaces  ...           30000.0
113839                             /organization/zengame-  ...               NaN
114946                                 /organization/eron  ...               NaN
114947                               /organization/asys-2  ...           18192.0
114948  /organization/novatiff-reklam-ve-tantm-hizmetl...  ...           14851.0

[74 rows x 6 columns]

This seems to work fine. 

Let's now look at the number of unique values in rounds dataframe again.

In [20]:
# Look at unique values again
len(rounds.company_permalink.unique())

66368

Now it makes sense - there are 66368 unique companies in both the ```rounds``` and ```companies``` dataframes. 

It is possible that a similar encoding problems are present in the companies file as well. Let's look at the companies which are present in the companies file but not in the rounds file - if these have special characters, then it is most likely because the ```companies``` file is encoded (while rounds is not).

In [21]:
# companies present in companies df but not in rounds df
companies.loc[~companies['permalink'].isin(rounds['company_permalink']), :]

permalink  ...  founded_at
43                               /organization/10â°north  ...  05-01-2013
426                  /organization/51wofang-无忧我房  ...  01-01-2014
1506                          /organization/adslinked™  ...         NaN
1775   /organization/aesthetic-everythingâ®-social-ne...  ...  15-06-2009
1834              /organization/affluent-attachã©-club-2  ...         NaN
...                                                  ...  ...         ...
63833                    /organization/whodat’s-spaces  ...         NaN
65778                 /organization/zengame-ç¦
游科暀  ...  17-07-2010
66365                               /organization/ãeron  ...  01-01-2011
66366                             /organization/ã”asys-2  ...  01-01-2014
66367  /organization/ä°novatiff-reklam-ve-tanä±tä±m-h...  ...         NaN

[68 rows x 10 columns]

Thus, the ```companies``` df also contains special characters. Let's treat those as well.

In [22]:
# remove encoding from companies df
companies['permalink'] = companies.permalink.str.encode('utf-8').str.decode('ascii', 'ignore')

Let's now look at the companies present in the companies df but not in rounds df - ideally there should be none.

In [23]:
# companies present in companies df but not in rounds df
companies.loc[~companies['permalink'].isin(rounds['company_permalink']), :]

Empty DataFrame
Columns: [permalink, name, homepage_url, category_list, status, country_code, state_code, region, city, founded_at]
Index: []

Thus, the encoding issue seems resolved now. Let's write these (clean) dataframes into separate files so we don't have to worry about encoding problems again.

In [24]:
# write rounds file
rounds.to_csv("rounds_clean.csv", sep=',', index=False)

# write companies file
companies.to_csv("companies_clean.csv", sep='\t', index=False)